nginx-sys/build.rs

```diff
@@ -196,34 +201,66 @@ build process:
    integrity of the downloaded files will not be verified.
 */
 
-fn zlib_archive_url() -> String {
-    let version = env::var("ZLIB_VERSION").unwrap_or_else(|_| ZLIB_DEFAULT_VERSION.to_string());
+fn zlib_archive_url(version: &String) -> String {
     format!("{ZLIB_DOWNLOAD_URL_PREFIX}/v{version}/zlib-{version}.tar.gz")
 }
 
-fn pcre2_archive_url() -> String {
-    let version = env::var("PCRE2_VERSION").unwrap_or_else(|_| PCRE2_DEFAULT_VERSION.to_string());
-    format!("{PCRE2_DOWNLOAD_URL_PREFIX}/pcre2-{version}/pcre2-{version}.tar.gz")
+fn pcre_archive_url(version: &String) -> String {
+    // We can distinguish pcre1/pcre2 by checking whether the second character is '.', because the final version of pcre1 is 8.45 and the first one of pcre2 is 10.00.
+    if version.chars().nth(1).is_some_and(|c| c == '.') {
+        format!("{PCRE1_DOWNLOAD_URL_PREFIX}/{version}/pcre-{version}.tar.gz")
+    } else {
+        format!("{PCRE2_DOWNLOAD_URL_PREFIX}/pcre2-{version}/pcre2-{version}.tar.gz")
+    }
 }
 
-fn openssl_archive_url() -> String {
-    let version = env::var("OPENSSL_VERSION").unwrap_or_else(|_| OPENSSL_DEFAULT_VERSION.to_string());
-    format!("{OPENSSL_DOWNLOAD_URL_PREFIX}/openssl-{version}/openssl-{version}.tar.gz")
+fn openssl_archive_url(version: &String) -> String {
+    if version.starts_with("1.1.1") {
+        let version_hyphened = version.replace(".", "_");
+        format!("{OPENSSL_DOWNLOAD_URL_PREFIX}/OpenSSL_{version_hyphened}/openssl-{version}.tar.gz")
+    } else {
+        format!("{OPENSSL_DOWNLOAD_URL_PREFIX}/openssl-{version}/openssl-{version}.tar.gz")
+    }
 }
 
-fn nginx_archive_url() -> String {
-    let version = env::var("NGX_VERSION").unwrap_or_else(|_| NGX_DEFAULT_VERSION.to_string());
+fn nginx_archive_url(version: &String) -> String {
     format!("{NGX_DOWNLOAD_URL_PREFIX}/nginx-{version}.tar.gz")
 }
 
 /// Returns a list of tuples containing the URL to a tarball archive and the GPG signature used
 /// to validate the integrity of the tarball.
 fn all_archives() -> Vec<(String, String)> {
+    let ngx_version = env::var("NGX_VERSION").unwrap_or_else(|_| NGX_DEFAULT_VERSION.into());
+    let zlib_version = env::var("ZLIB_VERSION").unwrap_or_else(|_| ZLIB_DEFAULT_VERSION.into());
+    // keep env name `PCRE2_VERSION` for compat
+    // Nginx had started supporting pcre2 and openssl3 since 1.22.0
```

Comment:
Could you please add a comment somewhere around here that restates what you wrote in the PR description:
> This PR enforces the build script of nginx-sys to download and use earlier versions of Nginx and its dependencies. This enables to build backports version targeted to LTS releases of Linux distributions. (ex. Ubuntu 22.04LTS provides a patched version of Nginx 1.18.0.)

This would provide context to someone reading this code and trying to understand why we had these logical branches around dependency versions.

Reply:
Added more explanations about the conditional branch of dependencies default version.